### PR TITLE
Vision 2024

### PR DESCRIPTION
**Pull request by spellingcat**
rewrote some stuff + added more details on pose est, 2024 setup, etc

### CODE REVIEW

**Comment by Lewis-Seiden** [comment]
making this merge to `2024` to match other 2024 prs

File: Docs/Specifics/Vision.md [hunk 1]
```diff
@@ -2,64 +2,113 @@
...
 
-Apriltags are a system of visual tags developed by researchers at the University of Michigan to provide low overhead, high accuracy localization for many different applications.
 They look similar to QR codes, but with less pixels.
-This trades amount of information in each tag for quick, easy, and robust detection.
-In the context of FRC, AprilTags are useful for helping your robot know where it is on the field, so it can align itself to some goal position.
-The idea behind this is that when the camera detects an Apriltag, it will calculate the distance, angle, etc. that the robot is from that target and then map that to a diagram that shows where all the Apriltags are on the field and then estimate where the robot is from there.
-This is useful because with a reliable pose estimation system we can automate more of the scoring process.
+This trades the amount of information in each tag for quick, easy, and robust detection.
+There are several different "families" of tags depending on how many bits a tag can represent.
+In 2023, the FRC game Charged Up used 16h5, while in 2024 (and presumably onwards) 36h11 was used instead.
```
Comment (Lewis-Seiden) [inline Docs/Specifics/Vision.md:21]: remove the "presumably onwards"

File: Docs/Specifics/Vision.md [hunk 2]
```diff
@@ -2,64 +2,113 @@
...
+![Robots at the University of Michigan with AprilTags](/Assets/Apriltags.jpg)
 
-Apriltags are a system of visual tags developed by researchers at the University of Michigan to provide low overhead, high accuracy localization for many different applications.
 They look similar to QR codes, but with less pixels.
-This trades amount of information in each tag for quick, easy, and robust detection.
-In the context of FRC, AprilTags are useful for helping your robot know where it is on the field, so it can align itself to some goal position.
-The idea behind this is that when the camera detects an Apriltag, it will calculate the distance, angle, etc. that the robot is from that target and then map that to a diagram that shows where all the Apriltags are on the field and then estimate where the robot is from there.
-This is useful because with a reliable pose estimation system we can automate more of the scoring process.
+This trades the amount of information in each tag for quick, easy, and robust detection.
+There are several different "families" of tags depending on how many bits a tag can represent.
```
Comment (Lewis-Seiden) [inline Docs/Specifics/Vision.md:20]: add a sentence about what the different types of families are for/how they differ. Something along the lines of "Different families have different sizes, shapes, and patterns of white, black, and "data" squares. This provides tradeoffs between detail, robustness, and number of potential IDs."

File: Docs/Specifics/Vision.md [hunk 3]
```diff
@@ -2,64 +2,113 @@
...
+In 2023, the FRC game Charged Up used 16h5, while in 2024 (and presumably onwards) 36h11 was used instead.
+In each frame in the camera stream, we are looking for these tags.
+Through a series of image processing algorithms (removing colors, decimation, identifying regions of relative lightness/darkness), we're able to identify both *which* tag it is (a numerical ID corresponding to the unique pattern of the tag) and *where* the corners of the tag are in the frame.
+
+In the context of FRC, because AprilTags are placed in various known locations around the field, we can use them in order for the robot to know where it is on the field.
+The simplest application of this is single tag detection and 2D alignment.
+For example, if we know that when a certain tag is in the middle of the camera's field of vision, our robot's shooter is lined up correctly with the target, we can write code that instructs the shooter to release a game piece once that tag has been centered, the drivetrain to adjust so that the tag is centered, etc.
```
Comment (Lewis-Seiden) [inline Docs/Specifics/Vision.md:27]: slightly long sentence, might want to rework

File: Docs/Specifics/Vision.md [hunk 4]
```diff
@@ -2,64 +2,113 @@
...
+How would an estimate be incorrect?
+One cause of this is pose ambiguity.
+
+![Demonstration of pose ambiguity](/Assets/Ambiguity.jfif)
+
+Humans are typically able to rely on things like depth perception and lighting to determine the difference between something like these two boxes, but it is more difficult for computers.
+If we also have pose estimates from other tags that are less ambiguous, we can throw out those (likely) incorrect estimates if they're significantly off.
+Having multiple cameras also helps with this, as another camera at a different angle could generate a more accurate pose with the same tag.
+Comparing a vision pose against odometry can also rule out incorrect poses.
+
```
Comment (Lewis-Seiden) [inline Docs/Specifics/Vision.md:51]: i would add a short paragraph explaining different disambiguation techniques and ambiguity, such as lowest ambiguity, closest to floor, and PNP.

File: Docs/Specifics/Vision.md [hunk 5]
```diff
@@ -2,64 +2,113 @@
...
+An important step to take before using 3D pose estimation is to correctly calibrate the camera.
+All camera lenses have some level of distortion that we need to account for when processing images gathered from them, especially in applications where it's important to accurately calculate where things are in 3D space.
+
+![Example of distorted chessboard](/Assets/Distortion.jpg)
+
+It's possible for us to undistort these images—we just need to find the camera-specific numbers that will allow us to perform that math, which we can do by calibrating the camera.
+There are a couple ways to do this, but one way is to take a lot of photos with the camera being calibrated of a chessboard pattern (which we know has straight lines and certain measurements) at different angles and positions and send those photos to programs that then handle the math for us and return those numbers.
+
```
Comment (Lewis-Seiden) [inline Docs/Specifics/Vision.md:59]: i would quickly cover that we can use built-in photonvision tools or mrcal for this

File: Docs/Specifics/Vision.md [hunk 6]
```diff
@@ -2,64 +2,113 @@
...
+### Object Detection
+
+Sometimes we want to be able to detect other things on the field that aren't AprilTags.
+This could include game pieces, other robots, and more.
```
Comment (Lewis-Seiden) [inline Docs/Specifics/Vision.md:63]: change to "This primarily includes game pieces, but can include other robots or more"

File: Docs/Specifics/Vision.md [hunk 7]
```diff
@@ -2,64 +2,113 @@
...
+One way of accomplishing this is through tuning HSV thresholds to only pick up on a certain color, such as the orange of a Crescendo note.
```
Comment (Lewis-Seiden) [inline Docs/Specifics/Vision.md:64]: explain what HSV means. may also be worth mentioning retro tape here

File: Docs/Specifics/Vision.md [hunk 8]
```diff
@@ -2,64 +2,113 @@
...
+Another way is contour filtering, which looks for the outlines of a shape like a circle or a triangle, such as a cone from Charged Up.
+
```
Comment (Lewis-Seiden) [inline Docs/Specifics/Vision.md:66]: include pictures of the game pieces?

File: Docs/Specifics/Vision.md [hunk 9]
```diff
@@ -2,64 +2,113 @@
...
+<img src="../../Assets/ShapeDetection.webp" alt="Example of detecting a triangular object" width="600"/>
+
+A more complicated system is through machine learning models that have been specially trained on the objects we want to detect, such as game pieces or other robots.
+Due to the significant processing power that this requires, in FRC this is typically done on the Rockchip Neural Processing Unit (NPU), which is on the Orange Pi 5 and similar coprocessors.
+While we haven't implemented this, the greater FRC community has developed several models and other resources.
+In 2024, Photonvision also ships with a pretrained model for notes.
```
Comment (Lewis-Seiden) [inline Docs/Specifics/Vision.md:72]: i would make more year agnostic such as "Photonvision has shipped ML based game piece detection recently, and likely will continue to."

Comment (Lewis-Seiden) [reply Docs/Specifics/Vision.md:72]: maybe include a link to the docs or smth idk

File: Docs/Specifics/Vision.md [hunk 10]
```diff
@@ -2,64 +2,113 @@
...
-It makes it a lot easier to interface with the camera and also has plenty of utility methods to get information like distance and position, as well as having some built in pose estimator features and more.
+Photonvision is a community, open source library that gets vision data from the robot for a variety of applications.
+It simplifies interfacing with the camera and also has plenty of utility methods to get information like distance and position, as well as having some built in pose estimator features and more.
 
 ### Limelight
 
 Limelight (LL) is a camera and processor system for FRC designed to make vision easy and simple.
-We used a Limelight 2+ with Limelight's software in 2022 for RR tape detection, and a LL2+ with Photonvision software in 2023 for Apriltag detection.
-The LL2+ was not satisfactory for apriltag detection for us, and we will likely move on to different hardware in the near future.
+We used a Limelight 2+ with Limelight's software in 2022 for retroreflective tape detection, and a LL2+ with Photonvision software in 2023 for Apriltag detection.
```
Comment (Lewis-Seiden) [inline Docs/Specifics/Vision.md:82]: add a sentence about LL3 and LL3G such as "Limelight has released the LL3 more recently which posses more processing power for Apriltags and object detection, and other teams have had success with it. The LL3G also features a global shutter camera for further improved Apriltag performance. At time of writing we have not tested either."

File: Docs/Specifics/Vision.md [hunk 11]
```diff
@@ -2,64 +2,113 @@
...
+The LL2+ was not satisfactory for Apriltag detection for us, so we switched to the...
+
+### Orange Pi + Arducam
+
+While we have the roboRIO to run much of our code on the robot, it is not capable of handling vision processing at a usable speed.
+Instead, we run this on a coprocessor, which is a second, faster computer also onboard the robot.
+For vision, we've been using the Orange Pi 5, a single board computer which is quite similar to Raspberry Pis you may have seen elsewhere.
+This is what Photonvision runs on.
```
Comment (Lewis-Seiden) [inline Docs/Specifics/Vision.md:90]: This is what we run Photonvision on.

File: Docs/Specifics/Vision.md [hunk 12]
```diff
@@ -2,64 +2,113 @@
...
+
+![Orange Pi 5](/Assets/OrangePi.jfif)
+
+Of course, we also need a camera.
+We've been using Arducam USB cameras, though other USB cameras also work.
```
Comment (Lewis-Seiden) [inline Docs/Specifics/Vision.md:95]: mention a few models we've used (and results w those) and any other common ones that we havent

File: Docs/Specifics/Vision.md [hunk 13]
```diff
@@ -2,64 +2,113 @@
...
-
-### Notes
+- [Photonlib examples](https://github.com/PhotonVision/photonvision/tree/master/photonlib-java-examples)
+- [8033 2024 implementation](https://github.com/HighlanderRobotics/Crescendo/tree/main/src/main/java/frc/robot/subsystems/vision)
 
-- Photonvision is installed on the coprocessor not on your computer, what you’re looking at on your computer is just a web dashboard
-- Last year we used a system called a Limelight which had a camera bundled in with a coprocessor. It had a couple of performance issues, so this year we’re probably going to use custom hardware **still tbd if we’re going to use an opi or not so i’ll come back and edit this**
 
-### Whats Next?
+### Exercises
```
Comment (Lewis-Seiden) [inline Docs/Specifics/Vision.md:112]: remove this section imo

**Review by Lewis-Seiden — CHANGES_REQUESTED** [review]

File: Docs/Specifics/Vision.md (hunk 9 above)
Comment (spellingcat) [reply Docs/Specifics/Vision.md:72]: yeah the docs are linked in the resources section

File: Docs/Specifics/Vision.md [hunk 14]
```diff
@@ -49,27 +51,40 @@ If we also have pose estimates from other tags that are less ambiguous, we can t
...
 Comparing a vision pose against odometry can also rule out incorrect poses.
 
+There are several different strategies to pick which pose we want to use.
+One of these is lowest ambiguity.
+This finds the pose corresponding to the vision result with the lowest reprojection error (a measurement of the difference between a measured 3D point and its projected 2D point in the camera frame).
+This is useful when we only have one tag in frame that we can use to differentiate between these possible poses.
+Another is to choose the pose closest to some other known pose, such as the camera's pose or the last calculated pose.
+Historically in FRC, it's been unlikely that a robot would be above the ground or outside the field, so this can eliminate such poses that are further than the robot could have traveled in the time since the last measurement.
+Lastly, the best way to take advantage of multiple tags is by using the multitag PNP (Perspective-n-Point) strategy.
+This maps the 3D points of visible AprilTags to 2D points to estimate the field-to-camera transformation.
```
Comment (Lewis-Seiden) [inline Docs/Specifics/Vision.md:61]: this doesnt really mean anything to a newbie, and isnt really a useful explanation. instead include something about how it fuses the poses from multiple tags

File: Docs/Specifics/Vision.md [hunk 15]
```diff
@@ -49,27 +51,40 @@ If we also have pose estimates from other tags that are less ambiguous, we can t
...
 All camera lenses have some level of distortion that we need to account for when processing images gathered from them, especially in applications where it's important to accurately calculate where things are in 3D space.
 
 ![Example of distorted chessboard](/Assets/Distortion.jpg)
 
 It's possible for us to undistort these images—we just need to find the camera-specific numbers that will allow us to perform that math, which we can do by calibrating the camera.
-There are a couple ways to do this, but one way is to take a lot of photos with the camera being calibrated of a chessboard pattern (which we know has straight lines and certain measurements) at different angles and positions and send those photos to programs that then handle the math for us and return those numbers.
+There are a couple ways to do this, but one way is to take a lot of photos with the camera being calibrated of a chessboard pattern (which we know has straight lines and certain measurements) at different angles and positions and send those photos to programs (such as PhotonVision or mrcal) that then handle the math for us and return those numbers.
```
Comment (Lewis-Seiden) [inline Docs/Specifics/Vision.md:69]: include links to relevant docs pages

**Review by Lewis-Seiden — CHANGES_REQUESTED** [review]
almost there

**Review by Lewis-Seiden — APPROVED** [review]

**Pull request merged**